src/classlibnative/bcltype/bignum.cpp

```diff
@@ -0,0 +1,608 @@
+    memset(result.m_blocks, 0, sizeof(UINT32) * BIGSIZE);
+
+    const UINT32* pLargeBegin = pLarge->m_blocks;
+    const UINT32* pLargeEnd = pLarge->m_blocks + pLarge->m_len;
+
+    UINT32* pResultStart = result.m_blocks;
+    const UINT32* pSmallCurrent = pSmall->m_blocks;
+    const UINT32* pSmallEnd = pSmallCurrent + pSmall->m_len;
+
+    while (pSmallCurrent != pSmallEnd)
+    {
+        // Multiply each block of large BigNum.
+        if (*pSmallCurrent != 0)
+        {
+            const UINT32* pLargeCurrent = pLargeBegin;
+            UINT32* pResultCurrent = pResultStart;
+            UINT64 carry = 0;
+
+            do
+            {
+                UINT64 product = (UINT64)(*pResultCurrent) + (UINT64)(*pSmallCurrent) * (UINT64)(*pLargeCurrent) + carry;
+                carry = product >> 32;
+                *pResultCurrent = (UINT32)(product & 0xFFFFFFFF);
+
+                ++pResultCurrent;
+                ++pLargeCurrent;
+            } while (pLargeCurrent != pLargeEnd);
+
+            *pResultCurrent = (UINT32)(carry & 0xFFFFFFFF);
+        }
+
+        ++pSmallCurrent;
+        ++pResultStart;
+    }
+
+    if (maxResultLength > 0 && result.m_blocks[maxResultLength - 1] == 0)
+    {
+        result.m_len = maxResultLength - 1;
+    }
+    else
+    {
+        result.m_len = maxResultLength;
+    }
+}
+
+void BigNum::Multiply10()
+{
+    if (IsZero())
+    {
+        return;
+    }
+
+    const UINT32* pCurrent = m_blocks;
+    const UINT32* pEnd = pCurrent + m_len;
+    UINT32* pResultCurrent = m_blocks;
+
+    UINT64 carry = 0;
+    while (pCurrent != pEnd)
+    {
+        UINT64 product = ((UINT64)(*pCurrent) << 3) +  ((UINT64)(*pCurrent) << 1) + carry;
+        carry = product >> 32;
+        *pResultCurrent = (UINT32)(product & 0xFFFFFFFF);
+
+        ++pResultCurrent;
+        ++pCurrent;
+    }
+
+    if (carry != 0)
+    {
+        _ASSERTE(m_len + 1 <= BIGSIZE);
+        *pResultCurrent = (UINT32)carry;
+        m_len += 1;
+    }
+}
+
+bool BigNum::IsZero() const
+{
+    return m_len == 0;
+}
+
+void BigNum::SetUInt32(UINT32 value)
+{
+    m_len = 1;
+    m_blocks[0] = value;
+}
+
+void BigNum::SetUInt64(UINT64 value)
+{
+    m_blocks[0] = (UINT32)(value & 0xFFFFFFFF);
+    m_blocks[1] = (UINT32)(value >> 32);
+    m_len = (m_blocks[1] == 0) ? 1 : 2;
+}
+
+void BigNum::SetZero()
+{
+    m_len = 0;
+}
+
+void BigNum::ExtendBlock(UINT32 blockValue)
+{
+    m_blocks[m_len] = blockValue;
+    ++m_len;
+}
+
+void BigNum::ExtendBlocks(UINT32 blockValue, UINT32 blockCount)
+{
+    _ASSERTE(blockCount > 0);
+
+    if (blockCount == 1)
+    {
+        ExtendBlock(blockValue);
+
+        return;
+    }
+
+    memset(m_blocks + m_len, 0, (blockCount - 1) * sizeof(UINT32));
+    m_len += blockCount;
+    m_blocks[m_len - 1] = blockValue;
+}
+
+UINT32 BigNum::LogBase2(UINT32 value)
+{
+    _ASSERTE(value != 0);
+
+#if defined(FEATURE_PAL)
+    return (UINT32) (8 * sizeof (UINT32) - __builtin_clz(value) - 1);
```

Comment:
Rather than using the `__builtin_clz` and `__buildtin_clzll` functions  here and having to have the ifdef, please add implementation of `_BitScanReverse` and `_BitScanReverse64` to pal.h instead. There are already implementations for the forward variants of the bit scan there, so please put these next to those.